File: src/output.ts
import type { NormalizedSnapshot } from "./types.js";

function padRight(s: string, width: number): string {
  if (s.length >= width) return s;
  return s + " ".repeat(width - s.length);
}

function clampText(s: string, width: number): string {
  if (s.length <= width) return s;
  if (width <= 1) return s.slice(0, width);
  return s.slice(0, width - 1) + "…";
}

function renderErrors(snapshot: NormalizedSnapshot): string[] {
  if (snapshot.errors.length === 0) return [];

  const lines: string[] = [];
  lines.push("");
  lines.push("Errors:");
  for (const e of snapshot.errors) {
    const action = e.actionable ? ` (next: ${e.actionable})` : "";
    lines.push(`- ${e.provider}: ${e.type}: ${e.message}${action}`);
  }
  return lines;
}

function makeBar(percent: number, width = 24): string {
  const p = Math.max(0, Math.min(100, Math.round(percent)));
  const filled = Math.round((p / 100) * width);
  return `${"█".repeat(filled)}${"░".repeat(width - filled)}`;
}

function prettyWindowTitle(provider: string, window: string): string {
  if (provider === "claude" && window === "claude:session") return "Claude Session limit";
  if (provider === "claude" && window === "claude:week(all-models)") {
    return "Claude Week limit";
  }
  if (provider === "codex" && window === "codex:5h") return "Codex Session limit";
  if (provider === "codex" && window === "codex:weekly") return "Codex Week limit";
  return window;
}

function prettyResetLabel(resetAt: string | null, notes: string | null): string {
  if (resetAt) {
    const ts = Date.parse(resetAt);
    if (Number.isFinite(ts)) {
      const formatted = new Intl.DateTimeFormat(undefined, {
        dateStyle: "medium",
        timeStyle: "short",
      }).format(new Date(ts));
      return `Resets ${formatted}`;
    }
  }
  if (notes) return `Resets ${notes.replace(/^Resets\s+/i, "")}`;
  return "Resets unknown";
}

export function renderGraph(snapshot: NormalizedSnapshot): string {
  const lines: string[] = [];
  lines.push(`Usage Snapshot @ ${snapshot.fetchedAt}`);
  lines.push("");

  if (snapshot.rows.length === 0) {
    lines.push("(no rows)");
  } else {
    for (const r of snapshot.rows) {
      const percent = r.usedPercent;
      const title = prettyWindowTitle(r.provider, r.window);
      const reset = prettyResetLabel(r.resetAt, r.notes);
      const meta = `${r.source}/${r.confidence}${r.stale ? "/stale" : ""}`;

      lines.push(title);
      if (percent != null) {
        lines.push(`  ${makeBar(percent, 32)}  ${Math.round(percent)}% used`);
      } else {
        const usageLabel =
          r.used != null && r.limit != null ? `${r.used}/${r.limit}` : "unknown";
        lines.push(`  ${usageLabel} used`);
      }
      lines.push(`  ${reset}`);
      lines.push(`  Source ${meta}`);
      lines.push("");
    }
    while (lines.length > 0 && lines[lines.length - 1] === "") lines.pop();
  }

  lines.push(...renderErrors(snapshot));
  return lines.join("\n");
}

export function renderTable(snapshot: NormalizedSnapshot): string {
  const rows = snapshot.rows.map((r) => {
    const usage =
      r.usedPercent != null
        ? `${Math.round(r.usedPercent)}%`
        : r.used != null && r.limit != null
          ? `${r.used}/${r.limit}`
          : "?";
    const reset = r.resetAt ?? "?";
    const meta = `${r.source}/${r.confidence}${r.stale ? "/stale" : ""}`;
    return {
      provider: r.provider,
      window: r.window,
      usage,
      reset,
      meta,
    };
  });

  const cols = {
    provider: 6,
    window: 16,
    usage: 10,
    reset: 25,
    meta: 18,
  };

  const header = [
    padRight("PROV", cols.provider),
    padRight("WINDOW", cols.window),
    padRight("USAGE", cols.usage),
    padRight("RESET_AT", cols.reset),
    padRight("META", cols.meta),
  ].join(" ");
  const sep = "-".repeat(header.length);

  const lines: string[] = [header, sep];
  if (rows.length === 0) {
    lines.push("(no rows)");
  } else {
    for (const r of rows) {
      lines.push(
        [
          padRight(clampText(r.provider, cols.provider), cols.provider),
          padRight(clampText(r.window, cols.window), cols.window),
          padRight(clampText(r.usage, cols.usage), cols.usage),
          padRight(clampText(r.reset, cols.reset), cols.reset),
          padRight(clampText(r.meta, cols.meta), cols.meta),
        ].join(" ")
      );
    }
  }

  lines.push(...renderErrors(snapshot));

  return lines.join("\n");
}

File: crates/qmeter-core/src/output.rs
use crate::types::{NormalizedError, NormalizedSnapshot};

fn pad_right(value: &str, width: usize) -> String {
    if value.len() >= width {
        return value.to_string();
    }
    format!("{value}{}", " ".repeat(width - value.len()))
}

fn clamp_text(value: &str, width: usize) -> String {
    if value.len() <= width {
        return value.to_string();
    }
    if width <= 1 {
        return value[..width].to_string();
    }
    format!("{}.", &value[..width - 1])
}

fn render_errors(snapshot: &NormalizedSnapshot) -> Vec<String> {
    if snapshot.errors.is_empty() {
        return Vec::new();
    }

    let mut lines = vec![String::new(), "Errors:".to_string()];
    for err in &snapshot.errors {
        lines.push(render_error(err));
    }
    lines
}

fn render_error(err: &NormalizedError) -> String {
    let action = err
        .actionable
        .as_ref()
        .map(|a| format!(" (next: {a})"))
        .unwrap_or_default();
    format!(
        "- {}: {:?}: {}{}",
        err.provider.as_str(),
        err.error_type,
        err.message,
        action
    )
}

fn usage_label(percent: Option<f64>, used: Option<u64>, limit: Option<u64>) -> String {
    if let Some(percent) = percent {
        return format!("{}%", percent.round() as i64);
    }
    match (used, limit) {
        (Some(used), Some(limit)) => format!("{used}/{limit}"),
        _ => "?".to_string(),
    }
}

fn pretty_window_title(provider: &str, window: &str) -> String {
    match (provider, window) {
        ("claude", "claude:session") => "Claude Session limit".to_string(),
        ("claude", "claude:week(all-models)") => "Claude Week limit".to_string(),
        ("codex", "codex:5h") => "Codex Session limit".to_string(),
        ("codex", "codex:weekly") => "Codex Week limit".to_string(),
        _ => window.to_string(),
    }
}

fn make_bar(percent: f64, width: usize) -> String {
    let bounded = percent.round().clamp(0.0, 100.0);
    let filled = ((bounded / 100.0) * width as f64).round() as usize;
    format!("{}{}", "#".repeat(filled), ".".repeat(width - filled))
}

pub fn render_graph(snapshot: &NormalizedSnapshot) -> String {
    let mut lines = vec![
        format!("Usage Snapshot @ {}", snapshot.fetched_at),
        String::new(),
    ];

    if snapshot.rows.is_empty() {
        lines.push("(no rows)".to_string());
    } else {
        for row in &snapshot.rows {
            let provider = row.provider.as_str();
            lines.push(pretty_window_title(provider, &row.window));
            if let Some(percent) = row.used_percent {
                let cache_tag = if row.source.is_cache() || row.stale {
                    " (cache)"
                } else {
                    ""
                };
                lines.push(format!(
                    "  {}  {}% used{}",
                    make_bar(percent, 32),
                    percent.round() as i64,
                    cache_tag
                ));
            } else {
                lines.push(format!(
                    "  {} used",
                    usage_label(row.used_percent, row.used, row.limit)
                ));
            }
            lines.push(format!(
                "  Resets {}",
                row.reset_at.as_deref().unwrap_or("unknown")
            ));
            lines.push(format!(
                "  Source {}/{}{}",
                row.source.as_str(),
                row.confidence.as_str(),
                if row.stale { "/stale" } else { "" }
            ));
            lines.push(String::new());
        }
        while lines.last().is_some_and(|line| line.is_empty()) {
            lines.pop();
        }
    }

    lines.extend(render_errors(snapshot));
    lines.join("\n")
}

pub fn render_table(snapshot: &NormalizedSnapshot) -> String {
    let cols = (6usize, 16usize, 18usize, 25usize, 18usize);
    let header = [
        pad_right("PROV", cols.0),
        pad_right("WINDOW", cols.1),
        pad_right("USAGE", cols.2),
        pad_right("RESET_AT", cols.3),
        pad_right("META", cols.4),
    ]
    .join(" ");
    let mut lines = vec![header.clone(), "-".repeat(header.len())];

    if snapshot.rows.is_empty() {
        lines.push("(no rows)".to_string());
    } else {
        for row in &snapshot.rows {
            let cache_tag = if row.source.is_cache() || row.stale {
                " (cache)"
            } else {
                ""
            };
            let usage = format!(
                "{}{}",
                usage_label(row.used_percent, row.used, row.limit),
                cache_tag
            );
            let meta = format!(
                "{}/{}{}",
                row.source.as_str(),
                row.confidence.as_str(),
                if row.stale { "/stale" } else { "" }
            );
            lines.push(
                [
                    pad_right(&clamp_text(row.provider.as_str(), cols.0), cols.0),
                    pad_right(&clamp_text(&row.window, cols.1), cols.1),
                    pad_right(&clamp_text(&usage, cols.2), cols.2),
                    pad_right(
                        &clamp_text(row.reset_at.as_deref().unwrap_or("?"), cols.3),
                        cols.3,
                    ),
                    pad_right(&clamp_text(&meta, cols.4), cols.4),
                ]
                .join(" "),
            );
        }
    }

    lines.extend(render_errors(snapshot));
    lines.join("\n")
}
